File: 127_word-ladder.cpp
#include "basicInclude.h"

//下面这个超时了...
class Solution {
public:
    bool ifDifferOne(string s1, string refer){
        if(s1.size() != refer.size()) return false;
        int differCount = 0;
        for(int i=0; i<s1.size(); i++){
            if(s1[i] != refer[i]){
                differCount++;
            }
            if(differCount > 1){
                return false;
            }
        }
        return differCount == 1;
    }
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        //如果结尾单词不在wordList中
        if(find(wordList.begin(), wordList.end(), endWord) == wordList.end()) return 0;
        queue<string> s1, s2;
        vector<bool> flags(wordList.size(), false);
        s1.push(beginWord);
        int ans = 0;
        while (!s1.empty() || !s2.empty())
        {
            ans++;
            if(s2.empty()){
                while(!s1.empty()){
                    string tmpStr = s1.front();
                    s1.pop();
                    for(int i=0; i<wordList.size(); i++){
                        if(ifDifferOne(tmpStr, wordList[i]) && !flags[i]){
                            if(wordList[i] == endWord){
                                return ans+1;
                            }
                            else {
                                flags[i] = true;
                                s2.push(wordList[i]);
                            }
                        }
                    }
                }
            }
            else {
                while(!s2.empty()){
                    string tmpStr = s2.front();
                    s2.pop();
                    for(int i=0; i<wordList.size(); i++){
                        if(ifDifferOne(tmpStr, wordList[i]) && !flags[i]){
                            if(wordList[i] == endWord){
                                return ans+1;
                            }
                            else {
                                flags[i] = true;
                                s1.push(wordList[i]);
                            }
                        }
                    }
                }
            }
        }
        return 0;
    }
};

class otherSolution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {

        unordered_set<string> dict(wordList.begin(), wordList.end());
        if (dict.find(endWord) == dict.end() ) return 0;
        // 初始化起始和终点
        unordered_set<string> beginSet, endSet, tmp, visited;
        beginSet.insert(beginWord);
        endSet.insert(endWord);
        int len = 1;

        while (!beginSet.empty() && !endSet.empty()){
            if (beginSet.size() > endSet.size()){
                tmp = beginSet;
                beginSet = endSet;
                endSet = tmp;
            }
            tmp.clear();
            for ( string word : beginSet){
                for (int i = 0; i < word.size(); i++){
                    char old = word[i];
                    for ( char c = 'a'; c <= 'z'; c++){
                        if ( old == c) continue;
                        word[i] = c;
                        if (endSet.find(word) != endSet.end()){
                            return len+1;
                        }
                        if (visited.find(word) == visited.end() && dict.find(word) != dict.end()){
                            tmp.insert(word);
                            visited.insert(word);
                        }
                    }
                    word[i] = old;
                }
            }
            beginSet = tmp;
            len++;
            

        }
        return 0;
    }
};

int main()
{
    string s1 = "hit";
    string s2 = "cog";
    vector<string> wordList = {"hot","dot","dog","lot","log","cog"};
    otherSolution s;
    cout<<s.ladderLength(s1, s2, wordList)<<endl;
    return 0;
}

File: 127_word-ladder.py
from typing import List
import collections

# BFS-Breadth first Search
class Solution:
    # 构造有向图
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        def addWord(word: str):
            if word not in wordId:
                nonlocal nodeNum
                wordId[word] = nodeNum
                nodeNum += 1
        
        def addEdge(word: str):
            addWord(word)
            id1 = wordId[word]
            chars = list(word)
            for i in range(len(chars)):
                tmp = chars[i]
                chars[i] = "*"
                newWord = "".join(chars)
                # 加入虚拟节点
                addWord(newWord)
                id2 = wordId[newWord]
                edge[id1].append(id2)
                edge[id2].append(id1)
                chars[i] = tmp

        # 记录节点与节点ID对应信息
        wordId = dict()
        # 记录边信息，邻接表
        edge = collections.defaultdict(list)
        nodeNum = 0

        for word in wordList:
            addEdge(word)
        addEdge(beginWord)

        if endWord not in wordId:
            return 0
        
        # 初始距离记录
        dis = [float("inf")] * nodeNum
        beginId, endId = wordId[beginWord], wordId[endWord]
        # 起点节点的距离为0
        dis[beginId] = 0

        # 从一个节点开始，广度优先搜索，并且维护距离数组dis
        que = collections.deque([beginId])
        while que:
            x = que.popleft()
            # 如果到达了自己想要查找的节点
            if x == endId:
                return dis[endId] // 2 + 1
            for it in edge[x]:
                # 如果没有遍历过
                if dis[it] == float("inf"):
                    dis[it] = dis[x] + 1
                    que.append(it)     
        return 0

class Test:
    def testFun(self):
        def test():
            nonlocal count
            print(count)
            count += 1
            # nonlocal dic
            dic.append(1)
            # print(count)

        dic = []
        count = 1
        test()
        print(dic)
        print(count)

if __name__ == "__main__":
    s = Solution()
    beginWord = "hit"
    endWord = "cog"
    wordList = ["hot","dot","dog","lot","log","cog"]
    print(s.ladderLength(beginWord, endWord, wordList))
